core/src/cluster_slots.rs

```diff
@@ -0,0 +1,267 @@
+use crate::{cluster_info::ClusterInfo, epoch_slots::EpochSlots, serve_repair::RepairType};
+
+use solana_ledger::{bank_forks::BankForks, staking_utils};
+use solana_sdk::{clock::Slot, pubkey::Pubkey};
+
+use std::{
+    collections::{HashMap, HashSet},
+    rc::Rc,
+    sync::RwLock,
+};
+
+#[derive(Default)]
+pub struct ClusterSlots {
+    cluster_slots: HashMap<Slot, HashMap<Rc<Pubkey>, u64>>,
+    keys: HashSet<Rc<Pubkey>>,
+    since: Option<u64>,
+    validator_stakes: HashMap<Rc<Pubkey>, u64>,
+    epoch: Option<u64>,
+    self_id: Pubkey,
+}
+
+impl ClusterSlots {
+    pub fn lookup(&self, slot: Slot) -> Option<&HashMap<Rc<Pubkey>, u64>> {
+        self.cluster_slots.get(&slot)
```

Comment:
FYI, in case it's ever needed, I have a function here that will do the mapping  from node pubkeys to their corresponding vote accounts:

https://github.com/solana-labs/solana/pull/8327/files#diff-c38da5b23679761df905a00ee4d2688dR50

but I'm not sure it's the best place for it. I should probably eventually add it to the bank and eat the work off the ABI change 😛